File: Year 3/Semester 1/PPD/lab01/java/Main.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    private static int N, M, n, m, p, offset;
    private static int[][] matrix, kernel, finalMatrix;

    public static void read(String path) {
        try {
            File myObj = new File(path);
            Scanner myReader = new Scanner(myObj);

            N = 10;
            M = 10;

            matrix = new int[N][M];
            if (myReader.hasNextLine()) {
                for (int i = 0; i < N; i++) {
                    String data = myReader.nextLine();
                    String[] line = data.split(" ");
                    for (int j = 0; j < M; j++) {
                        matrix[i][j] = Integer.parseInt(line[j]);
                    }
                }
            }

            n = 3;
            m = 3;

            kernel = new int[n][m];
            if (myReader.hasNextLine()) {
                for (int i = 0; i < n; i++) {
                    String data = myReader.nextLine();
                    String[] line = data.split(" ");
                    for (int j = 0; j < m; j++) {
                        kernel[i][j] = Integer.parseInt(line[j]);
                    }
                }
            }
            myReader.close();
        } catch (FileNotFoundException e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        }
    }

    public static void write(String path, int[][] matrix) {
        try {
            BufferedWriter bw = new BufferedWriter(new FileWriter(path));
            for (int[] elem : matrix) {
                for (int i : elem) {
                    bw.write(i + " ");
                }
                bw.newLine();
            }
            bw.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static boolean checkAnswer() throws Exception {
        File fileTest = new File("output.txt");
        Scanner scannerTest = new Scanner(fileTest);

        File fileAnswer = new File("answer0.txt");
        Scanner scannerAnswer = new Scanner(fileAnswer);

        while (scannerTest.hasNextLine() && scannerAnswer.hasNextLine()) {
            String test = scannerTest.nextLine();
            String answer = scannerAnswer.nextLine();
            test = test.substring(0, test.length() - 1);
            System.out.println(test + "\n" + answer);
            if (!answer.equals(test)) {
                throw new Exception("Invalid output");
            }
        }

        if (scannerTest.hasNextLine() || scannerAnswer.hasNextLine()) {
            throw new Exception("Invalid output");
        }

        return true;
    }

    public static int Convolution(int x, int y) {
        int output = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                int ii = x - offset + i;
                int jj = y - offset + j;

                if (ii < 0) ii = 0;
                else if (ii >= N) ii = N - 1;
                if (jj < 0) jj = 0;
                else if (jj >= M) jj = M - 1;

                output += matrix[ii][jj] * kernel[i][j];
            }
        }
        System.out.println(output);
        return output;
    }

    public static void sequentialConvolution() {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                finalMatrix[i][j] = Convolution(i, j);
            }
        }
    }

    public static void parallelConvolution() throws InterruptedException {
        Thread[] t = new Worker[p];

        int start, end = 0, chunk, rest;
        if (N > M) {
            chunk = N / p;
            rest = N % p;
        } else {
            chunk = M / p;
            rest = M % p;
        }

        long startTime = System.nanoTime();

        for (int i = 0; i < t.length; i++) {
            start = end;
            end = start + chunk;
            if (rest > 0) {
                end++;
                rest--;
            }
            t[i] = new Worker(M, start, end, finalMatrix);
            t[i].start();
        }

        for (Thread thread : t) {
            thread.join();
        }

        long stopTime = System.nanoTime();

        System.out.println((double) (stopTime - startTime) / 1E6);
    }

    public static class Worker extends Thread {
        int M, start, end;
        int[][] finalMatrix;

        public Worker(int M, int start, int end, int[][] finalMatrix) {
            this.M = M;
            this.start = start;
            this.end = end;
            this.finalMatrix = finalMatrix;
        }

        public void run() {
            if (N > M) {
                for (int i = start; i < end; i++) {
                    for (int j = 0; j < M; j++) {
                        this.finalMatrix[i][j] = Main.Convolution(i, j);
                    }
                }
            } else {
                for (int i = 0; i < N; i++) {
                    for (int j = start; j < end; j++) {
                        this.finalMatrix[i][j] = Main.Convolution(i, j);
                    }
                }
            }

        }
    }

    public static void blockDistr(List<Pair<Integer, Integer>> blocks, int start, int end) {
        System.out.println(start + " " +  end);
        for (int b = start; b < end; b++) {
            for (int i = blocks.get(b).first; i < blocks.get(b).first + 5; i++) {
                for (int j = blocks.get(b).second; j < blocks.get(b).second + 5; j++) {
                    System.out.println("si aici");
                    finalMatrix[i][j] = Convolution(i, j);
                }
            }
        }
    }

    public static void blockConvolution() {
        Thread[] threads = new Thread[p];
        List<Pair<Integer, Integer>> blocks = new ArrayList<>();
        for (int i = 0; i < N; i += 5) {
            for (int j = 0; j < M; j += 5) {
                blocks.add(new Pair<>(i, j));
            }
        }

        int len = blocks.size();
        int cat = len / p;
        int rest = len % p;
        int last = 0;
        for (int i = 0; i < p; i++) {
            int start = last;
            int end = start + cat;
            if (i < rest) {
                start += 1;
            }
            last = end;
            int finalStart = start;
            threads[i] = new Thread(() -> blockDistr(blocks, finalStart, end));
            threads[i].start();
        }

        for (int i = 0; i < p; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public static Pair<Integer, Integer> getPosition(int index) {
        int row = index / M;
        int col = index % M;
        return new Pair<>(row, col);
    }

    public static void linearDistr(int startIndex, int endIndex) {
        for (int i = startIndex; i < endIndex; i++) {
            Pair<Integer, Integer> position = getPosition(i);
            int ii = position.first;
            int jj = position.second;
            finalMatrix[ii][jj] = Convolution(ii, jj);
        }
    }

    public static void linearConvolution() {
        Thread[] threads = new Thread[p];
        int cat = (N * M) / p;
        int rest = (N * M) % p;
        int last = 0;
        for (int i = 0; i < p; i++) {
            int start = last;
            int end = start + cat;
            if (i < rest) {
                end += 1;
            }
            last = end;
            int finalEnd = end;
            threads[i] = new Thread(() -> linearDistr(start, finalEnd));
            threads[i].start();
        }

        for (int i = 0; i < p; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public static void jumpingDistr(int index, int jump) {
        for (int i = index; i < N * M; i += jump) {
            Pair<Integer, Integer> position = getPosition(i);
            int ii = position.first;
            int jj = position.second;
            finalMatrix[ii][jj] = Convolution(ii, jj);
        }
    }

    public static void jumpingConvolution() {
        Thread[] threads = new Thread[p];
        for (int i = 0; i < p; i++) {
            int finalI = i;
            threads[i] = new Thread(() -> jumpingDistr(finalI, p));
            threads[i].start();
        }

        for (int i = 0; i < p; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }


    public static void main(String[] args) throws Exception {
        read("data0.txt");
        finalMatrix = new int[N][M];

        p = 2;

        offset = (n - 1) / 2;
        System.out.println(p);
        long startTime = System.nanoTime();

        if (p == 1)
            sequentialConvolution();
        else {
            //parallelConvolution();
            //blockConvolution();
            //linearConvolution();
            jumpingConvolution();
        }

        long endTime = System.nanoTime();
        double difference = (endTime - startTime) / 1E6;
        System.out.println(difference);

        write("output.txt", finalMatrix);

        if (checkAnswer() == false) {
            System.out.println("Tests failed\n");
        } else {
            System.out.println("Tests passed\n");
        }
    }
}

class Pair<A, B> {
    public A first;
    public B second;

    public Pair(A first, B second) {
        this.first = first;
        this.second = second;
    }
}

File: Year 3/Semester 1/PPD/lab01/c/main.cpp
#include <iostream>
#include <fstream>
#include <chrono>
#include <thread>

#define N 10000
#define M 10
#define n 5
#define m 5
#define p 16

int matrix[N][M];
int kernel[n][m];
int finalMatrix[N][M];
int answerMatrix[N][M];

// vector<vector<int>> matrix;
// vector<vector<int>> kernel;
// vector<vector<int>> finalMatrix;
// vector<vector<int>> answerMatrix;

using namespace std;
using namespace std::chrono;

ifstream fin("../data3.txt");
ofstream fout("../output.txt");

void read() {

    // kernel = vector<std::vector<int>>(n, std::vector<int>(m));
    // matrix = vector<std::vector<int>>(N, std::vector<int>(M));
    // finalMatrix = vector<std::vector<int>>(N, std::vector<int>(M));

    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            fin >> kernel[i][j];

    for (int i = 0; i < N; i++)
        for (int j = 0; j < M; j++) {
            fin >> matrix[i][j];
        }
}

void readAnswer() {
    ifstream fin_ans("../answer3.txt");

    for (int i = 0; i < N; i++)
        for (int j = 0; j < M; j++) {
            fin_ans >> answerMatrix[i][j];
        }
}

void write() {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++)
            fout << finalMatrix[i][j] << " ";
        fout << endl;
    }

    fout.close();
}

int Convolution(int x, int y) {
    int output = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            int ii = x - (n - 1) / 2 + i;
            int jj = y - (m - 1) / 2 + j;

            if (ii < 0) ii = 0;
            else if (ii >= N) ii = N - 1;

            if (jj < 0) jj = 0;
            else if (jj >= M) jj = M - 1;
            output += matrix[ii][jj] * kernel[i][j];
        }
    }
    return output;
}

void secventialConvolution() {

    for (int i = 0; i < N; i++)
        for (int j = 0; j < M; j++)
            finalMatrix[i][j] = Convolution(i, j);
}

void parallelDistr(int start, int end) {
    if (N > M) // mai multe linii decat coloane
    {
        for (int i = start; i < end; i++)
            for (int j = 0; j < M; j++)
                finalMatrix[i][j] = Convolution(i, j);
    } else // mai multe coloane decat linii
    {
        for (int i = 0; i < N; i++)
            for (int j = start; j < end; j++)
                finalMatrix[i][j] = Convolution(i, j);
    }
}

void parallelConvolution() {
    vector<thread> t;

    int start, end = 0, chunk, rest;
    if (N > M) {
        chunk = N / p;
        rest = N % p;
    } else {
        chunk = M / p;
        rest = M % p;
    }

    for (size_t i = 0; i < p; i++) {
        start = end;
        end = start + chunk;
        if (rest > 0) {
            end++;
            rest--;
        }
        thread thr = thread(parallelDistr, start, end);
        t.push_back(std::move(thr));
        start = end;
    }

    for (auto &th: t) {
        if (th.joinable())
            th.join();
    }
}

bool checkAnswer() {
    readAnswer();

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            if (finalMatrix[i][j] != answerMatrix[i][j])
                return 0;
        }
    }
    return 1;
}

void blockDistr(vector<pair<int, int>> blocks, int start, int end) {
    for (int b = start; b < end; b++) {
        for (int i = blocks[b].first; i < blocks[b].first + 5; i++) {
            for (int j = blocks[b].second; j < blocks[b].second + 5; j++) {
                finalMatrix[i][j] = Convolution(i, j);
            }
        }
    }
}

void blockConvolution() {
    vector<thread> t;
    vector<pair<int, int>> blocks;
    for (int i = 0; i < N; i += 5) {
        for (int j = 0; j < M; j += 5) {
            blocks.push_back({i, j});
        }
    }

    int len = blocks.size();
    int cat = len / p;
    int rest = len % p;
    int last = 0;
    for (int i = 0; i < p; i++) {
        int start = last;
        int end = start + cat;
        if (i < rest) {
            start += 1;
        }
        last = end;
        thread thr = thread(blockDistr, blocks, start, end);
        t.push_back(std::move(thr));
    }


    for (auto &th: t) {
        if (th.joinable())
            th.join();
    }
}

pair<int, int> getPosition(int index) {
    int row = index / M;
    int col = index % M;

    return {row, col};
}

void linearDistr(int start, int end) {
    for (int i = start; i < end; i++) {
        pair<int, int> position = getPosition(i);
        int ii = position.first;
        int jj = position.second;
        finalMatrix[ii][jj] = Convolution(ii, jj);
    }
}

void linearConvolution() {
    vector<thread> t;
    int cat = (N * M) / p;
    int rest = (N * M) % p;
    int last = 0;
    for (int i = 0; i < p; i++) {
        int start = last;
        int end = start + cat;
        if (i < rest) {
            end += 1;
        }
        last = end;
        thread thr = thread(linearDistr, start, end);
        t.push_back(std::move(thr));
    }

    for (auto &th: t) {
        if (th.joinable())
            th.join();
    }
}

void jumpingDistr(int index, int jump) {
    for (int i = index; i < N * M; i += jump) {
        pair<int, int> position = getPosition(i);
        int ii = position.first;
        int jj = position.second;
        finalMatrix[ii][jj] = Convolution(ii, jj);
    }
}

void jumpingConvolution() {
    vector<thread> t;
    for (int i = 0; i < p; i++) {
        thread thr = thread(jumpingDistr, i, p);
        t.push_back(std::move(thr));
    }

    for (int i = 0; i < p; i++) {
        for (auto &th: t) {
            if (th.joinable())
                th.join();
        }
    }
}

int main() {
    read();

    auto startTime = high_resolution_clock::now();

    if (p == 1)
        secventialConvolution();
    else {
        //parallelConvolution();
        //blockConvolution();
        //linearConvolution();
        jumpingConvolution();
    }

    auto endTime = high_resolution_clock::now();

    double difference = duration<double, milli>(endTime - startTime).count();

    cout << difference;

    write();

    if (!checkAnswer())
        //    cout << "Tests passed!\n";
        //else
        cout << "Tests failed!\n";

    return 0;
}
